### ISSUE TITLE
`Type[]` ignored in `Type[Callable[P, R]` -- maybe bug?

### ISSUE THREAD

**Issue by supersergiy** [microsoft/pyright#4178]
The following snippet ([playground link](https://pyright-playground.decorator-factory.su/?gzip=H4sIAPxWbWMC_6WRzW6DMAzH734Kq5dCxXgApE6bJu2Mtmo7VFXkBVOhBoKctBtvvxAKnHaaL7Fj_3_yRy22RT_0TXfGpu2teDwMPWf4QsbQlwneczdkWIr1VluTxfQHCdSrUvGP5841tnMzpCSh9r1nDfCmtMX9rEs2Y7zJUNsbSUOd3x_kyimUoWZRJdtymwJoQ87hp1Dfc5XMPRzLDEfIKS0Ag1Vco1I6NKxU4tjUGe5Izq7AMh_fEO4u3_PP5KX48BghE2K0PM9h4bV04f-yRk7F2gp5K0nMaRN04yaO836XYU4QOTF5n3jNTeRIfVqQ9_W8WvvnHqhA5yWCwxnX_oT9VTokAOEbk1HhkJwEUrqUAPwCTRuPURwCAAA%3D)):
```python
from typing import Type, Callable, Any, Protocol, TypeVar
from typing_extensions import ParamSpec

R_co = TypeVar("R_co", covariant=True)
P = ParamSpec('P')

class Wrapped(Protocol[P, R_co]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
        ...

def decorator(
    cls: Type[Callable[P, R_co]]
) -> Type[Wrapped[P, R_co]]:
    ...

@decorator
class Foo:
    def __call__(self, a: str) -> Any:
        return a

reveal_type(Foo)
```

Reveals: `Type of "Foo" is "Type[Wrapped[(*args: Any, **kwargs: Any), Foo]]"`

This makes me confused. The `[(*args: Any, **kwargs: Any), Foo]` suggests that the `__call__` of object creation is used instead of the object's `__call__`, which is what I would expect when using `cls: Callable[P, R_co]`, but not `cls: Type[Callable[P, R_co]]`. To me it seems like Pyright ignores the `Type[]` here -- omitting `Type[]` results in the same revealed type as the above ([playground link](https://pyright-playground.decorator-factory.su/?gzip=H4sIAC9abWMC_6WRzWrDMAzH73oK0Uud4uUBAh0bg53DVrZDKUZzlRLqxEF2u-Xt5yRNetppuliy9P-hj0p8g7Hv6vaEddN5ibjrO9b4Qs7Rl0vec9trLMVHb73TY_qDBKq70vBP5DbUvg0zpCSh5r1jC_BmrMftrFOrIV5ptP5KUlMbtzu5cAZlqllUal2uMwDrKAT8FOo6Pqq5h32pcYAcsgIw2ZErNMamho1RgV2lcUNyCgWW-fCmcHP-nn8mL8OHxxEyIQbL8xwWXkNn_i9r4BzZeqHoRY0565JuXu0yB4yEYUH726xL6jAxR97TArst5tX7PzdABYYoIzgd8N6ZcLxIiwQgfGVyJp2QVSJlSwnAL1Q6NDYWAgAA)).

The solution here seems to be using a custom callable protocol that specifies `self` as the first argument ([playground link](https://pyright-playground.decorator-factory.su/?gzip=H4sIAA1YbWMC_72SzWrDMAyA734K0UuT4uUBAh0bg90GYSvboRSjuUoJdeIgu93y9nOcOjntNJgulizp0w-q2bbgh77pTtC0vWUPu6EnCU9oDH6aoD12g4SKrbfaGhnd78iiXjIVfXvqXGM7lyAVMrZvPWkhXpW2sE152Wq0VxK0vSI32Pntji-UiyrEzFnZulrnQmiDzsHLkJrJUhv7SsLIOeSlgCBHqkEpHcKUyhyZWsIG-eRKqIrxDebm_JV-Ji2Hu_sImRCjFEWRin4w9j0d_6Vi4rV4pr-yRs6RtGX0lrPo0ybkjevfL5ucxzmISIru28yLb2JH7sMMvS3o2dpfN4ElOM8RHK5n6ZDJX7gDFILpSmhUuB_KAimfQ4T4AR6Tl6yTAgAA)):

```python
from typing import Type, Callable, Any, Protocol, TypeVar
from typing_extensions import ParamSpec

R_co = TypeVar("R_co", covariant=True)
P = ParamSpec('P')

class MyCallable(Protocol[P, R_co]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
        ...

class Wrapped(Protocol[P, R_co]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R_co:
        ...

def decorator(
    cls: Type[MyCallable[P, R_co]]
) -> Type[Wrapped[P, R_co]]:
    ...

@decorator
class Foo:
    def __call__(self, a: str) -> Any:
        return a

reveal_type(Foo)
```
Which reveals the intuitive: `Type of "Foo" is "Type[Wrapped[(a: str), Any]]"`

I'm really curious why it works this way. Is ignoring `Type[]` around `typing.Callable` a bug or a feature? Why does using a custom protocol with `self` prevent it from being ignored?

Thank you

**Comment by erictraut** [comment]
`Type[X]` only works with classes. It means "the class that you need to instantiate to create an object of type `X`". 

`Callable` is a special form that has no instantiable counterpart.

If your intent is to describe a class that has a `__call__` method, you should use a protocol as you have done above.

Classes themselves are callable (because the metaclass `type` supports `__call__` to implement class instantiation), so there's no need to use `Type` in a decorator parameter. You can simply use:

```python
def decorator(cls: Callable[P, R_co]) -> Wrapped[P, R_co]:
    ...
```

**Comment by supersergiy** [comment]
In this case, shouldn't a warning be raised on `Type[Callable]`? Many users might not expect that `Callable` would behave differently from a callable protocol here. I couldn't find anything about it in the [docs](https://docs.python.org/3/library/typing.html#typing.Callable).

**Comment by erictraut** [comment]
Maybe. This is the first time I've seen anyone attempt to use `Type[Callable]`. It didn't even occur to me that anyone would ever try this. It's definitely not something that comes up often.

**Comment by supersergiy** [comment]
Thanks for the explanations, got it.

For the record, I got there because I needed a way to add a function to a user-defined class which implements `__call__`, and I needed the signature of the new function to match the signature of `__call__`. It seemed easier to preserve all the types with a decorator than inheritance. I switched to `pyright` here because `mypy` doesn't work with class decorators without plugings, and even with plugins it's a bit wonky. I think is a nice advantage of `pyright`.
